controllers/PostsController.js

```diff
@@ -22,22 +22,89 @@ class PostsController {
+    let { sortBy, order, limit, page } = req.query;
+    const orderList = ["ASC", "DESC"];
+    try {
+      if (!!limit && isNaN(Number(limit))) {
+        throw new Error("Wrong type of limit");
+      }
+      if (!!page && isNaN(Number(page))) {
+        throw new Error("Wrong type of page");
+      }
+      if (!!page && !limit) {
+        throw new Error("Must have limit for page");
+      }
+      if (!!order && !orderList.includes(order)) {
+        throw new Error(`order: Wrong value of order ("ASC"/"DESC")`);
+      }
+      if (!!sortBy && !order) {
+        order = "DESC";
+      }
+      const postsOption = this.getPostsOption(sortBy, order, limit, page);
+      const categoryInstance = await this.category.findOne({
+        where: { name: category },
+      });
+      if (!categoryInstance) {
+        throw new Error(`Category: ${category} cannot be found `);
+      }
+      const posts = await categoryInstance.getPosts({
+        joinTableAttributes: [],
+        ...postsOption,
+      });
+      return res.json(posts);
+    } catch (error) {
+      return res.status(400).send(error.message);
     }
-    if (!data.title || !data.content) {
-      return res.status(400).send("Must have title/content data");
+  }
+
+  getPostsOption(sortBy, order, limit, page) {
+    const postsOption = { include: [{ model: this.user, as: "author" }] };
+    if (!!limit && !page) {
+      page = 1;
     }
-    if (typeof data.title !== "string" || typeof data.content !== "string") {
-      return res.status(400).send("Wrong Type of title/content");
+    if (!!limit) {
+      const offset = (page - 1) * limit;
+      postsOption.offset = offset;
+      postsOption.limit = limit;
     }
-    if (!data.title.length || !data.content.length) {
-      return res.status(400).send("Must have content for title/content");
+    switch (sortBy) {
+      case "newestPost":
+        postsOption.order = [["createdAt", order]];
+        break;
+      case "newestComment":
+        break;
+      case "popular":
+        break;
+      case undefined:
+        break;
+      default:
+        throw new Error(
+          `ValueError: Wrong value of sortBy ("newestPost"/"newestComment"/"popular")`
+        );
     }
+    return postsOption;
+  }
+
+  async createPost(req, res) {
+    const { categories, ...data } = req.body;
     const t = await this.sequelize.transaction();
     try {
+      if (isNaN(Number(data.authorId))) {
+        throw new Error("Wrong Type of authorId");
+      }
+      if (!data.title || !data.content) {
+        throw new Error("Must have title/content data");
+      }
+      if (typeof data.title !== "string" || typeof data.content !== "string") {
+        throw new Error("Wrong Type of title/content");
+      }
+      if (!data.title.length || !data.content.length) {
+        throw new Error("Must have content for title/content");
+      }
+
       const newPost = await this.post.create(data, { transaction: t });
```

Comment:
What does transaction do?

Reply:
Transaction is a record of manipulation of the database. It helps when a single request will do multiple manipulation. If one of the manipulations goes wrong. It will undo all the handling by rollback to prevent broken data.